src/background/utils/requests.js

```diff
@@ -188,58 +191,36 @@ function decodeBody(obj) {
 // });
 
 // Modifications on headers
-{
-  function onBeforeSendHeaders(details) {
-    const headers = details.requestHeaders;
-    let newHeaders = [];
-    const vmHeaders = {};
-    headers.forEach((header) => {
-      // if (header.name === 'VM-Task') {
-      //   tasks[details.requestId] = header.value;
-      // } else
-      if (header.name.startsWith('VM-')) {
-        vmHeaders[header.name.slice(3)] = header.value;
-      } else {
-        newHeaders.push(header);
-      }
-    });
-    const reqId = vmHeaders.Verify;
-    if (reqId) {
-      const req = requests[reqId];
-      if (req) {
-        delete vmHeaders.Verify;
-        verify[details.requestId] = reqId;
-        req.coreId = details.requestId;
-        Object.keys(vmHeaders).forEach((name) => {
-          if (isSpecialHeader(name.toLowerCase())) {
-            newHeaders.push({ name, value: vmHeaders[name] });
-          }
-        });
-        if (req.anonymous) {
-          // Drop cookie in anonymous mode
-          newHeaders = newHeaders.filter(({ name }) => name.toLowerCase() !== 'cookie');
-        }
-      }
-    }
-    return { requestHeaders: newHeaders };
-  }
-  const filter = {
-    urls: ['<all_urls>'],
-    types: ['xmlhttprequest'],
+const transformedHeaders = Object.create(null);
+function onBeforeSendHeaders({ requestHeaders: headers, requestId }) {
+  const reqId = (headers.find(h => h.name === VM_VERIFY) || {}).value;
+  const req = reqId && requests[reqId];
+  if (!reqId || !req) return { requestHeaders: headers };
+  verify[requestId] = reqId;
+  req.coreId = requestId;
+  // if the listener is to be unregistered, it must occur in the next event loop cycle
+  setTimeout(transformHeaders, 0, reqId);
```

Comment:
Hmm, I wonder if onBeforeSendHeaders may be called multiple times during redirects? If so I need to unregister later, probably by using onResponseStarted and/or a bigger timeout.